src/kv/RocksDBStore.cc

```diff
@@ -3051,74 +3038,65 @@ int RocksDBStore::reshard(const std::string& new_sharding, const RocksDBStore::r
       //check if need to write batch
       if (bytes_in_batch >= ctrl.bytes_per_batch ||
 	  keys_in_batch >= ctrl.keys_per_batch) {
-	flush_batch();
+	flush_batch(&bat);
 	if (ctrl.unittest_fail_after_first_batch) {
-	  r = -1000;
-	  goto out;
+	  return -1000;
 	}
       }
     }
-    flush_batch();
-  out:
-    delete it;
-    delete bat;
-    return r;
+    if (bat.Count() > 0) {
+      flush_batch(&bat);
+    }
+    return 0;
   };
 
-  r = prepare_for_reshard(new_sharding, to_process_columns, to_process_handles);
+  auto close_column_handles = make_scope_guard([this] {
+    cf_handles.clear();
+    close();
+  });
+  columns_t to_process_columns;
+  int r = prepare_for_reshard(new_sharding, to_process_columns);
   if (r != 0) {
     dout(1) << "failed to prepare db for reshard" << dendl;
-    goto cleanup;
+    return r;
   }
 
-  ceph_assert(to_process_columns.size() == to_process_handles.size());
-  for (size_t idx = 0; idx < to_process_columns.size(); idx++) {
-    dout(5) << "Processing column=" << to_process_columns[idx] <<
-      " handle=" << to_process_handles[idx] << dendl;
-    if (to_process_columns[idx] == rocksdb::kDefaultColumnFamilyName) {
-      ceph_assert(to_process_handles[idx] == default_cf);
+  for (auto& [name, handle] : to_process_columns) {
+    dout(5) << "Processing column=" << name
+	    << " handle=" << handle.get() << dendl;
+    if (name == rocksdb::kDefaultColumnFamilyName) {
+      ceph_assert(handle.get() == default_cf);
       r = process_column(default_cf, std::string());
     } else {
-      std::string fixed_prefix = to_process_columns[idx].substr(0, to_process_columns[idx].find('-'));
+      std::string fixed_prefix = name.substr(0, name.find('-'));
       dout(10) << "Prefix: " << fixed_prefix << dendl;
-      r = process_column(to_process_handles[idx], fixed_prefix);
+      r = process_column(handle.get(), fixed_prefix);
     }
     if (r != 0) {
-      derr << "Error processing column " << to_process_columns[idx] << dendl;
-      goto cleanup;
+      derr << "Error processing column " << name << dendl;
+      return r;
     }
     if (ctrl.unittest_fail_after_processing_column) {
-      r = -1001;
-      goto cleanup;
+      return -1001;
     }
   }
 
-  r = reshard_cleanup(to_process_columns, to_process_handles);
+  r = reshard_cleanup(to_process_columns);
   if (r != 0) {
     dout(5) << "failed to cleanup after reshard" << dendl;
-    goto cleanup;
+    return r;
   }
 
   if (ctrl.unittest_fail_after_successful_processing) {
-    r = -1002;
-    goto cleanup;
+    return -1002;
   }
   env->CreateDir(sharding_def_dir);
-  status = rocksdb::WriteStringToFile(env, new_sharding,
-				      sharding_def_file, true);
-  if (!status.ok()) {
+  if (auto status = rocksdb::WriteStringToFile(env, new_sharding,
+					       sharding_def_file, true);
+      !status.ok()) {
```

Comment:
Since `status` is local, maybe just `!rocksdb::Write...(...).ok()` ?

Reply:
@aclamk thanks for your review Adam! i will implement your suggestion in a follow up PR. as some rados suites are failing due to this issue, i am going to merge it as it is.